store/postgres/src/mapping.rs

```diff
@@ -0,0 +1,1141 @@
+pub struct Table {
+    /// The name of the GraphQL object type ('Thing')
+    pub object: s::Name,
+    /// The name of the database table for this type ('things')
+    pub name: SqlName,
+    /// The name for a single object of elements of this type in GraphQL
+    /// queries ('thing')
+    pub singular_name: String,
+
+    pub columns: Vec<Column>,
+    /// The position of this table in all the tables for this mapping; this
+    /// is really only needed for the tests to make the names of indexes
+    /// predictable
+    position: u32,
+}
+
+impl Table {
+    fn new(
+        defn: &s::ObjectType,
+        interfaces: &mut HashMap<String, Vec<SqlName>>,
+        id_type: IdType,
+        position: u32,
+    ) -> Result<Table, StoreError> {
+        let table_name = Table::collection_name(&defn.name);
+        let columns = defn
+            .fields
+            .iter()
+            .map(|field| Column::new(field, id_type))
+            .collect::<Result<Vec<_>, _>>()?;
+        let table = Table {
+            object: defn.name.clone(),
+            name: table_name.clone(),
+            singular_name: Table::object_name(&defn.name),
+            columns,
+            position,
+        };
+        for interface_name in &defn.implements_interfaces {
+            match interfaces.get_mut(interface_name) {
+                Some(tables) => tables.push(table.name.clone()),
+                None => {
+                    return Err(StoreError::Unknown(format_err!(
+                        "unknown interface {}",
+                        interface_name
+                    )))
+                }
+            }
+        }
+        Ok(table)
+    }
+    /// Find the column `name` in this table. The name must be in snake case,
+    /// i.e., use SQL conventions
+    pub fn column(&self, name: &SqlName) -> Result<&Column, StoreError> {
+        self.columns
+            .iter()
+            .find(|column| &column.name == name)
+            .ok_or_else(|| StoreError::UnknownField(name.to_string()))
+    }
+
+    /// Find the column for `field` in this table. The name must be the
+    /// GraphQL name of an entity field
+    pub fn column_for_field(&self, field: &str) -> Result<&Column, StoreError> {
+        self.columns
+            .iter()
+            .find(|column| &column.field == field)
+            .ok_or_else(|| StoreError::UnknownField(field.to_string()))
+    }
+
+    pub fn primary_key(&self) -> Reference {
+        Reference {
+            table: self.name.clone(),
+            column: PRIMARY_KEY_COLUMN.into(),
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn reference(&self, name: &SqlName) -> Result<Reference, StoreError> {
+        let attr = self.column(name)?;
+        Ok(Reference {
+            table: self.name.clone(),
+            column: attr.name.clone(),
+        })
+    }
+
+    pub fn collection_name(gql_type_name: &str) -> SqlName {
+        SqlName::from(gql_type_name.to_snake_case().to_plural())
```

Comment:
For the sake of simplicity, lets not use `to_plural()` here.

Reply:
Commit 68ebd99d